### PR TITLE
Encryption added to the backup

### PR DESCRIPTION
**Pull request by oscarporsius**
Added GPG encryption to the docker-backup.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -112,6 +112,10 @@ skipped. Also when uploading the backup using SFTP goes wrong, the backups will
 the container. The Ofelia logs will also be written to `/backup/ofelia` so they remain persistent even when re-creating
 the container.
 
+## Encryption
+The backups can be encrypted with [GPG](https://www.gnupg.org/). To use this feature, you need to provide a GPG key 
+The GPG keys are used to encrypt and decrypt the backup files. You need to provide a "Public key file in .asc format" and place it in the keyfile folder. Make sure you name the file `public-key.asc`. Also make sure you rebuil the image with the new key. The private key is not. You need the private key to decrypt the backups. __make sure you keep the private key in a safe place. If you loose the private key that accompanies the public key, you will not be able to decrypt the backups!__
```
Comment (mprins) [inline README.md:117]:
```suggestion
The GPG keys are used to encrypt and decrypt the backup files. You need to provide a "Public key file in .asc format" and place it in the keyfile folder. Make sure you name the file `public-key.asc`. Also make sure you rebuild the image with the new key. The private key is not. You need the private key to decrypt the backups. __make sure you keep the private key in a safe place. If you loose the private key that accompanies the public key, you will not be able to decrypt the backups!__ 
```

Comment (mprins) [inline README.md:117]: > The private key is not.

Not what?

Comment (oscarporsius) [reply README.md]: Should have been "The private key is not supposed to be used with this container!"

File: Dockerfile [hunk 2]
```diff
@@ -23,13 +23,14 @@ RUN apt-get update && \
...
 RUN apt-get install -y -q --no-install-recommends bash postgresql-client-${POSTGRES_CLIENT_VERSION} && \
-    apt-get autoremove -yqq --purge wget ca-certificates gnupg2 && \
+    apt-get autoremove -yqq --purge wget ca-certificates && \
     rm -rf /var/lib/apt/lists/*
 
 WORKDIR /home/backup
 COPY *.sh .
 RUN chmod +x *.sh
 COPY include include
+COPY keyfile/public-key.asc include
```
Comment (matthijsln) [inline Dockerfile:33]: The public key should not be included in Docker image, but be a runtime environment variable. The script can write the environment variable to a file before calling `gpg`.

A public key has newlines but it shouldn't be a problem to configure a multiline environment variable using yaml or `\n` in an env-file. If multiline environment variables are still problematic the script can do some text mangling.

File: include/encrypt.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        return
+    fi
+    encrypt_dir=${DIR_BACKUP}
+    success=true  # assume success initially
+
+    start_time=$(date +"%s")   # record the current time in seconds since epoch
+
+    for file in "${encrypt_dir}"/*; do
+        print_msg "encrypting file: ${file} "
+        gpg --quiet --output "${file}.gpg" --encrypt --recipient-file "${KEYFILE}" "${file}"
```
Comment (matthijsln) [inline include/encrypt.sh:15]: If performance is slow `xargs` with the `-P` option can be used to parallellize encryption using multiple cores.

**Review by matthijsln — CHANGES_REQUESTED** [review]
See comments.

File: include/utils.sh [hunk 4]
```diff
@@ -20,4 +20,16 @@ upload_backup() {
   rm $DIR_UPLOADED/* 2>/dev/null
   mkdir -p ~/.ssh && ssh-keyscan -H $SFTP_HOST 2> /dev/null > $HOME/.ssh/known_hosts
   sshpass -e scp -p $DIR_BACKUP/* $SFTP && mv $DIR_BACKUP/* $DIR_UPLOADED/ || print_error "[ERR] Error copying to SFTP server"
+}
+
+upload_encrypted_backup() {
```
[the remark this answers is not in the document]
Comment (matthijsln) [reply include/utils.sh:25]: If leaving the unencrypted files on the host is intended for convenience, the duplicate code can still be reduced.

File: .env [hunk 5]
```diff
@@ -0,0 +1,5 @@
+# This environment file is used to set up a Public GPG key for use in the encryption proces.
```
Comment (matthijsln) [inline .env:1]: Just remove this file

File: include/encrypt.sh [hunk 6]
```diff
@@ -8,10 +8,10 @@ encrypt_files() {
     mkdir -p /home/backup/include/keyfile
```
Comment (matthijsln) [inline include/encrypt.sh:8]: Use `$(dirname $KEYFILE)` or `${KEYFILE%/*}`

File: test-compose.yml [hunk 7]
```diff
@@ -0,0 +1,36 @@
+# Add this to your Docker Compose configuration, or merge it by referencing this file in the COMPOSE_FILE environment
```
Comment (matthijsln) [inline test-compose.yml:1]: Remove

File: README.md [hunk 8]
```diff
@@ -150,6 +149,42 @@ This container is configured using the following environment variables:
...
 the compression as is the case with other compression tools (even the parallel versions).
 
+**Exporting a GPG for use in an environment file.**
+=====================================================
+
+You can use the below command to export the fingerprint of a specific GPG key, and remove unnecessary information.
+
+```gpg --armor --export [key-name] | tail -n +3 | head -n -1 | tr -d "[:space:]"```
+
+### Step-by-Step Explanation:
```
Comment (matthijsln) [inline README.md:159]: Remove

**Review by matthijsln — CHANGES_REQUESTED** [review]

**Review by oscarporsius — COMMENTED** [review]
lgtm

**Review by oscarporsius — COMMENTED** [review]
lgtm

**Review by oscarporsius — COMMENTED** [review]
lgtm

**Pull request merged**